docs/serialization.md

```diff
@@ -0,0 +1,98 @@
+
+If possible, avoid using reflection-based serializers with trimming, and prefer solutions based on source generators where the serialized types and all required members are statically referenced.
+
+As a last resort, the linker does have limited heuristics that can be enabled to keep _some_ of the types and members required for serialization, but this provides no correctness guarantees; apps which use reflection-based serializers are still considered "broken" as far as the static analysis can tell, and it is up to you to make sure that the app works as intended.
+
+## History
+
+The linker has historically been used for Xamarin scenarios that use reflection-based serializers like XmlSerializer, since before the introduction of the trim analysis warnings. There were limited heuristics to satisfy some simple uses of serializers. To provide backwards compatibility for such scenarios, the linker has built-in heuristics that makes some simple cases "just work", albeit in an opaque and unpredictable way.
+
+Consider disabling this behavior by passing `--disable-serialization-discovery` if possible, but it may be necessary when using legacy serializers that don't provide source generators or a similar solution that is statically analyzable. The following is a description of the heuristics for anyone who is unfortunate enough to have to rely on this behavior.
+
+## Heuristics
+
+Serialization discovery is enabled by default, and can be disabled by passing `--disable-serialization-discovery`. There are three parts to the heuristics:
+- Root type discovery: logic to discover types which are considered for serialization
+- Type graph: recursive logic to build a set of types to consider for serialization, starting from the roots
+- Preservation logic: what the linker does with the discovered types
+
+## Root type discovery
+
+The heuristics will discover types that satisfy _all_ of the following criteria:
+
+- The type is used
+
+  There must be a statically discoverable reference to the type. In other words, if running the linker without the serialization heuristics removes a given type, then the heuristics will not keep it either.
+
+- The type or one of its members must be attributed with a serializer-specific attribute.
+
+  See the sections below about the attributes you can use for each serializer. The attribute must be present on the type, or one of its defined members, including fields/properties/methods/events, public or private, though the serializers may not define attributes that can be placed on all member kinds. It is not enough to place an attribute on a base type or on a member defined by a base type.
+
+Note that passing a type directly to a serializer constructor is _not_ enough to keep it. We do not use dataflow to discover types. For example:
+
+```csharp
+new XmlSerializer (typeof (RootType)); // Will not keep RootType
```

Comment:
```suggestion
new XmlSerializer (typeof (RootType)); // Will not consider RootType as one of the root types
```

Linker will preserve the RootType since it's referenced.